pipeline/inferrer/inference_manager/src/main/scala/uk/ac/wellcome/platform/inference_manager/services/InferenceManagerWorkerService.scala

```diff
@@ -38,48 +50,94 @@ class InferenceManagerWorkerService[Destination](
+    FlowWithContext[DownloadedImage, Ctx].mapConcat { image =>
+      inferrerAdapters.map { inferrerAdapter =>
+        (
+          inferrerAdapter.createRequest(image),
+          (image, inferrerAdapter)
+        )
+      }
     }
 
   private def unmarshalResponse[Ctx] =
-    FlowWithContext[(Try[HttpResponse], DownloadedImage), Ctx]
+    FlowWithContext[
+      (Try[HttpResponse], (DownloadedImage, InferrerAdapter)),
+      Ctx]
       .map {
-        case result @ (_, image) =>
+        case result @ (_, (image, _)) =>
           imageDownloader.delete.runWith(Source.single(image))
           result
       }
       .mapAsync(parallelism) {
-        case (Success(response), image) =>
-          inferrerAdapter
+        case (Success(response), (image, adapter)) =>
+          adapter
             .parseResponse(response)
+            .map(Success(_))
             .recover {
               case e: Exception =>
                 response.discardEntityBytes()
-                throw e
+                Failure(e)
+            }
+            .map { response =>
+              AdapterResponseBundle(
+                image,
+                adapter,
+                response
+              )
             }
-            .map((image, _))
         case (Failure(exception), _) =>
           Future.failed(exception)
       }
 
-  private def augmentInput[Ctx] =
-    FlowWithContext[
-      (DownloadedImage, Option[inferrerAdapter.InferrerResponse]),
-      Ctx]
-      .map {
-        case (image, response) =>
-          inferrerAdapter.augmentInput(image, response)
+  private def collectAndAugment[Ctx <: Message] =
+    FlowWithContext[AdapterResponseBundle[DownloadedImage], Ctx]
+      .via {
+        Flow[(AdapterResponseBundle[DownloadedImage], Ctx)]
+          .groupBy(parallelism, _ match {
+            case (_, msg) => msg.messageId()
+          }, allowClosedSubstreamRecreation = true)
+          .groupedWithin(inferrerAdapters.size, maxInferrerWait)
+          .map { elements =>
+            elements.filter {
+              case (AdapterResponseBundle(_, _, Failure(_)), _) => false
+              case _                                            => true
+            }
+          }
+          .map { elements =>
+            if (elements.size != inferrerAdapters.size) {
+              val failedInferrers =
+                inferrerAdapters.map(_.getClass.getSimpleName) --
+                  elements.map(_._1.adapter.getClass.getSimpleName).toSet
+              throw new Exception(
+                s"Did not receive responses from $failedInferrers within $maxInferrerWait")
+            }
+            elements
+          }
+          .map { elements =>
+            val inferredData = elements.foldLeft(InferredData.empty) {
+              case (
+                  data,
+                  (AdapterResponseBundle(_, adapter, Success(response)), _)) =>
+                adapter.augment(data, response.asInstanceOf[adapter.Response])
```

Comment:
Ah in light of my comment above a parameterised type would be best for the `Try` argument so we can avoid a type cast here.